disk/test/com/treode/disk/edit/SimpleLogSpec.scala

```diff
@@ -0,0 +1,98 @@
+package com.treode.disk
+
+import java.nio.file.{Paths, StandardOpenOption}
+import com.treode.async.{Async, Scheduler}
+import com.treode.async.io.File
+import com.treode.async.io.stubs.StubFile
+import com.treode.async.stubs.StubScheduler
+import scala.util.{Failure, Success}
+import com.treode.buffer.PagedBuffer
+import com.treode.async.Async
+import com.treode.async.Async.async
+import org.scalatest._
+import com.treode.async.implicits._
+import com.treode.async.stubs.implicits._
+import com.treode.async.Globals
+
+class SimpleLog (file: StubFile, geom: DriveGeometry) (implicit scheduler: Scheduler) {
+   import geom.{blockAlignUp, blockAlignDown}
+   
+   var pos = 0L
+   val buf = PagedBuffer (10)
+   
+   def record (s: String): Async [Unit] =
+   async { cb =>
+      // read pos is aligned at start, and after each record
+      // write pos is at end of last record
+      if (buf.writePos > 0) {
+         buf.writePos -= 1
+         buf.writeByte(1)
+      }
+      buf.writeString (s)
+      buf.writeByte(0)
+      
+      val end = buf.writePos // remember where we parked
+      buf.writePos = geom.blockAlignUp (buf.writePos)
+      
+      file.flush (buf, pos) run {
+         case Success (length) => {
+            // flush move readPos to == writePos
+            buf.readPos = geom.blockAlignDown (end)
+            buf.writePos = end
+            pos += buf.discard (buf.readPos)
+            cb.pass()
+         }
+         case Failure (thrown) => cb.fail (thrown)
+      }
+      
+   }
+}
+
+class SimpleLogSpec extends FlatSpec {
+   import com.treode.disk.DriveGeometry
+   
+   "SimpleLog" should "record once to a StubFile" in {
+      implicit val scheduler = StubScheduler.random()
+      var testfile = StubFile(new Array[Byte](50), 0)
+      val rec = new SimpleLog(testfile, DriveGeometry(10, 10, 16384))
+      var str = "hithere"
+      rec.record(str).expectPass()
+      
+      var input = PagedBuffer(12)
+      testfile.fill (input, 0, str.length()) .expectPass()
+      
+      var actual = input.readString()
```

Comment:
This is strange:

    var x = in.readString()
    x += in.readByte()
    assert (x == "string"+0)

You can just say:

    assert (in.readString() == "string")
    assert (in.readByte() == 0)